Add rebuild support when setting a child online/offline

paulyoong opened the pull request:
- Setting a child online kicks off a rebuild job for the destination child
- Setting a child offline stops any existing rebuilds for the destination child

Additional changes:
- Fix the "child_operation" call to support online and offline operations
- Make the "stop_rebuild" call idempotent
- Fix the "online" and "offline" rpc calls so that they do not return a result

Commit:
Add example rebuild operations to mctl (#2)

Add unimplemented mctl operations for getting rebuild state and rebuild progress.
This will serve as a reference for adding future rebuild operations to mctl.

tiagolobocastro committed:
Simple rebuild cas 37 36 (#1)

Adding a simple/noddy rebuild and its test

Rebuild up to 10MiB of the child to avoid timing out the js tests: this is temporary until we move the rebuild trigger to another thread context

Add a test case for the rebuild within the add_child rpc call:
1 - create/start a nexus with 1 child
2 - dd from urandom into the nexus
3 - verify (cmp) that the child has the right data
5 - verify (cmp) that a new child (to be added) has the wrong data
4 - add the new child and wait for rebuild
5 - verify (cmp) that the new child has the right data

CAS-37: On child added fully rebuild from another healthy child on the same context with no frontend IO
CAS-36: Develop the first test case for the child added rebuild process

tiagolobocastro committed:
delete test disk files if they already exist other we'd fail the test (#3)

tiagolobocastro committed:
Add initial CI build configuration file (#6)

Add custom GitHub Actions file that builds and tests using the generic test.sh file.

I've created "special" build container: teamrebuild/ci to speed up the build from 14 to 8 minutes, otherwise we can just use the mayadata/ms-buildenv:latest container...

FROM mayadata/ms-buildenv:latest
WORKDIR /usr/src/app
# nix
COPY shell.nix .
COPY nix ./nix
# js tests
COPY mayastor-test/*.json ./mayastor-test/
RUN nix-shell --run 'cd mayastor-test; npm install'
# rust
COPY vendor ./vendor
COPY config .

The vendor cache is generated using the command "cargo vendor > .cache/config" from your nix-shell and then adjusting the vendor path to /usr/src/app

I've added a docker image to: https://github.com/teamrebuild/teamrebuild-gen-ci

We have to disable the dangling NBD test as it seems to fail all the time, suspect it's because we're running on quite an older kernel on azure...

tiagolobocastro committed:
Check code style and common programming errors in JS and rust source code (#7)

and Format previous simple rebuild code according to cargo clippy

tiagolobocastro committed:
Fix js prettier and enable prettier test for CI (#10)

Fix js prettier and enable prettier test for CI

tiagolobocastro committed:
CAS-94 Implement rebuild task as scoped in CAS-96

When a child is added it is put in a faulted state. A "start_rebuild" rpc shall bring it up to date by copying data from any healthy child (which is then pinned to the rebuild task). Upon completion the nexus is notified so that it may set the child to Open and reconfigure the io channels allowing the child to be integral part of the nexus. (no host IO during rebuild assumed)

coded by: paulyoong & tiagolobocastro

tiagolobocastro committed:
Merge remote-tracking branch 'openebs/master'

# Conflicts:
#	mayastor/src/bdev/nexus/nexus_bdev_children.rs
#	mayastor/src/bdev/nexus/nexus_rpc.rs
#	mayastor/src/rebuild.rs
#	mayastor/tests/nexus_rebuild.rs

tiagolobocastro committed:
Merge remote-tracking branch 'OpenEbs/master'

tiagolobocastro committed:
CI issues as of late, breaks down the tests for future debugging

Commit:
Add support to stop rebuild (#14)

Add support to stop a rebuild task

Additional changes:
- Add support to get the rebuild state
- Include the total number of rebuilds in the nexus information

Commit:
Merge upstream master into branch (#15)

Bring fork up-to-date with upstream.

tiagolobocastro committed:
rebuild: Move rebuild task list from the nexus into a global scope (#149)

This makes it easier to interact with the rebuild task list using its destination alone as a key
Also tidy up some of the doc comments and reorder the code a bit

also bring back CI yaml file since we've rebased master off of openebs/master

tiagolobocastro committed:
rebuild: Move rebuild task list from the nexus into a global scope #16

This makes it easier to interact with the rebuild task list using its destination alone as a key
Also tidy up some of the doc comments and reorder the code a bit

also bring back CI yaml file since we've rebased master off of openebs/master

Commit:
Add pause/resume rebuild support

Commit:
Add pause/resume rebuild support

Commit:
Simplify error handling

Commit:
Simplify error handling

Commit:
Merge branch 'pause-resume-rebuild' of https://github.com/teamrebuild/Mayastor into pause-resume-rebuild

Commit:
Merge master into branch

Commit:
Add pause/resume unit tests

Commit:
Undo unnecessary merge changes

Commit:
Address review comments

Commit:
Revert nix file

Commit:
Use latest nix file from upstream

Commit:
Start rebuild when bringing child online

Commit:
Add pause/resume rebuild support (#158)

tiagolobocastro committed:
Rename RebuildTask to RebuildJob and move rebuild code...

... to a separate file to facilitate mainline rebases
(plus we can always change it back when we integrate into master)

Commit:
Add offline/online unit tests

Additional changes:
- Fix the "child_operation" call to support online and offline operations
- Make the "stop_rebuild" call idempotent
- Fix the "online" and "offline" rpc calls so that do not return a result

Commit:
Merge master into branch

Commit:
Fixup merge changes

tiagolobocastro commented on csi/src/mayastor_svc.rs:
```diff
@@ -307,8 +307,21 @@ impl service::mayastor_server::Mayastor for MayastorService {
         trace!("{:?}", msg);
-        jsonrpc::call::<_, ()>(&self.socket, "offline_child", Some(msg))
-            .await?;
+        match msg.action {
```
Could we add a helper child action to string? And maybe we return an error is it's an invalid action?

tiagolobocastro commented on mayastor/src/bdev/nexus/nexus_rpc.rs:
```diff
@@ -167,21 +167,29 @@ pub(crate) fn register_rpc_methods() {
-    jsonrpc_register("offline_child", |args: ChildNexusRequest| {
-        let fut = async move {
-            let nexus = nexus_lookup(&args.uuid)?;
-            nexus.offline_child(&args.uri).await
```
why are we not returning the state now?

paulyoong replied:
When calling the offline/online rpc methods, the rpc clients were complaining that they were receiving a return value (the nexus state) when they weren't expecting to get anything back (they were expecting a unit type back which is () or empty). If you look at the child_operation rpc implementation in mayastor_svc.rs we were doing something similar i.e. ignoring the returned value and returning null.

Commit:
Address review comments

Commit:
Keep child in open state when set online

Commit:
Add child as faulted when set online

Additional changes:
- Add a separate function to wait for the rebuild job to complete

tiagolobocastro reviewed (approved):
reviewed the new changes on a call so LGTM
